File: mario-cpp/core/src/enemy.cpp
#include "mario/core/enemy.hpp"

#include <algorithm>

#include "mario/core/physics.hpp"
#include "mario/core/types.hpp"

namespace mario::core {

void Enemy::reset(Vec2 spawn_tile, const World& world, const Config& config) {
  size = config.enemy_size;
  const units_t tile = config.tile_size;
  const units_t x = spawn_tile.x + (tile - size.x) / 2;
  const units_t sample_x = spawn_tile.x + tile / 2;
  const units_t base_y = world.ground_y_for_x(sample_x, spawn_tile.y, config).value_or(spawn_tile.y + tile);
  const units_t y = base_y - size.y;

  pos = Vec2{x, y};
  vel = Vec2{0, 0};
  dir = -1;
  alive = true;
  on_ground = false;
}

void Enemy::update(const World& world, const Config& config) {
  if (!alive) {
    return;
  }

  vel.y = std::min(vel.y + config.gravity, config.terminal_velocity);
  vel.x = static_cast<units_t>(dir) * config.enemy_speed;

  const units_t desired_x = vel.x;
  const physics::MoveResult moved = physics::move_with_collisions(pos, size, vel, world.solids);
  const bool hit_wall = (desired_x != 0) && (moved.vel.x == 0);

  pos = moved.pos;
  vel = moved.vel;
  on_ground = moved.on_ground;

  if (hit_wall) {
    dir *= -1;
    vel.x = static_cast<units_t>(dir) * config.enemy_speed;
  } else if (on_ground) {
    const units_t foot_x =
        (dir >= 0) ? (pos.x + size.x + px_to_units(1)) : (pos.x - px_to_units(1));
    const units_t foot_y = pos.y + size.y + px_to_units(1);

    bool has_ground = false;
    if (const auto ground_y = world.ground_y_for_x(foot_x, foot_y, config); ground_y.has_value()) {
      has_ground = (*ground_y <= foot_y);
    }

    if (!has_ground) {
      dir *= -1;
      vel.x = static_cast<units_t>(dir) * config.enemy_speed;
    }
  }

  const units_t world_w = static_cast<units_t>(world.width) * config.tile_size;
  if (pos.x <= 0) {
    pos.x = 0;
    dir = 1;
  } else if (pos.x + size.x >= world_w) {
    pos.x = std::max<units_t>(0, world_w - size.x);
    dir = -1;
  }
}

}  // namespace mario::core

File: mario-rust/src/game/enemy.rs
use macroquad::prelude::*;

use super::{physics, sprites::Sprites, world::World, Config};

#[derive(Clone)]
pub struct Enemy {
    pub pos: Vec2,
    pub vel: Vec2,
    dir: f32,
    pub alive: bool,
    size: Vec2,
    on_ground: bool,
}

impl Enemy {
    pub fn new(tile_pos: Vec2, world: &World, config: &Config) -> Self {
        let size = config.enemy_size;
        let tile = config.tile_size;
        let x = tile_pos.x + (tile - size.x) * 0.5;
        let sample_x = tile_pos.x + tile * 0.5;
        let base_y = world
            .ground_y_for_x(sample_x, tile_pos.y, config)
            .unwrap_or(tile_pos.y + tile);
        let y = base_y - size.y;

        Self {
            pos: vec2(x, y),
            vel: Vec2::ZERO,
            dir: -1.0,
            alive: true,
            size,
            on_ground: false,
        }
    }

    pub fn reset(&mut self, tile_pos: Vec2, world: &World, config: &Config) {
        *self = Self::new(tile_pos, world, config);
    }

    pub fn update(&mut self, world: &World, config: &Config, dt: f32) {
        if !self.alive {
            return;
        }

        self.vel.y = (self.vel.y + config.gravity * dt).min(config.terminal_velocity);
        self.vel.x = config.enemy_speed * self.dir;

        let desired_x = self.vel.x;
        let (pos, vel, on_ground) =
            physics::move_with_collisions(self.pos, self.size, self.vel, &world.solids, dt);

        let hit_wall = desired_x.abs() > f32::EPSILON && vel.x.abs() <= f32::EPSILON;
        self.pos = pos;
        self.vel = vel;
        self.on_ground = on_ground;

        if hit_wall {
            self.dir *= -1.0;
            self.vel.x = config.enemy_speed * self.dir;
        } else if self.on_ground {
            let foot_x = if self.dir >= 0.0 {
                self.pos.x + self.size.x + 1.0
            } else {
                self.pos.x - 1.0
            };
            let foot_y = self.pos.y + self.size.y + 1.0;
            if world
                .ground_y_for_x(foot_x, foot_y, config)
                .map(|ground_y| ground_y <= foot_y)
                != Some(true)
            {
                self.dir *= -1.0;
                self.vel.x = config.enemy_speed * self.dir;
            }
        }

        let world_w = world.width as f32 * config.tile_size;
        if self.pos.x <= 0.0 {
            self.pos.x = 0.0;
            self.dir = 1.0;
        } else if self.pos.x + self.size.x >= world_w {
            self.pos.x = (world_w - self.size.x).max(0.0);
            self.dir = -1.0;
        }
    }

    pub fn rect(&self) -> Rect {
        physics::rect_at(self.pos, self.size)
    }

    pub fn draw(&self, sprites: &Sprites) {
        if !self.alive {
            return;
        }

        draw_texture_ex(
            sprites.chestnut_guy(),
            self.pos.x,
            self.pos.y,
            WHITE,
            DrawTextureParams {
                dest_size: Some(self.size),
                flip_x: self.vel.x < 0.0,
                ..Default::default()
            },
        );
    }
}
